doc/developers/cython.rst

```diff
@@ -0,0 +1,178 @@
+* tips to develop Cython code in scikit-learn
+
+Cython Internals
+----------------
+
+Concrete method dispatch implementation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* In most cases, concrete method dispatch is performed **at runtime using virtual tables**
+  (i.e. static C-array of pointers to a given types of function, used at runtime to known which
+  function to call from an instance). This is costly because there is a double indirection.
+
+* In some cases, concrete method dispatch is performed **statically** (a function is called
+  directly by a pointer dereference). This is less costly, especially if there are rooms
+  for this function to be inline.
+
+Known Limitations
+^^^^^^^^^^^^^^^^^
+
+* No template for types exist in Cython, alternatives are:
+
+  * using fused-types
+  * using Tempita, a small preprocessing language to expand Cython source
+
+* Multiple inheritance is not possible for extension types
+* No concept similar to interfaces or traits exist in Cython
+* Currently it is impossible to inherit and override a method which has a fused type in its signature
+* Structs cannot be created using constructor syntax with `nogil` `issue and workaround <https://github.com/cython/cython/issues/1642>`_
+
+Performances improvements will come with `Cython 3.0 which is to be released soon in beta apparently <https://github.com/cython/cython/issues/4022#issuecomment-1340918590>`_
+
+Basics
+^^^^^^
+
+* ``nogil`` declaration: https://github.com/cython/cython/issues/2798#issuecomment-459971828
+
+Tips for developping Cython code in scikit-learn
+------------------------------------------------
+
+Tips to ease development
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+* If you intend to use OpenMP: On MacOS, system's distribution of ``clang`` does not implement OpenMP.
+  You can install the ``compilers`` package available on ``conda-forge`` which comes with implementations of OpenMP.
+
+* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:
+
+  .. code-block:: bash
+
+         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1
+
+* Start from scratch in a notebook to understand.
+* To debug C code (e.g. a segfault), do use ``gdb`` with:
+
+  .. code-block:: bash
+
+         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py
+
+* To have access to some value in place to debug in ``cdef (nogil)`` context, use:
+
+  .. code-block:: cython
+
+         with gil:
+             print(state_to_print)
+
+* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.
+    ``print(f"{test_val=}")``
+
+* scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and we need to simplified
+  and unify that across the codebase. For now, make sure you understand which concrete types are used ultimately.
```

Comment:
you can maybe add a link to the corresponding issue where the ongoing work is tracked https://github.com/scikit-learn/scikit-learn/issues/25572
```suggestion
* scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and there's ongoing work to simplify
  and unify that across the codebase. For now, make sure you understand which concrete types are used ultimately.
```